LET'S PLAY WITH **DATA

In [3]:
import pandas as pd

In [12]:
import pandas as pd
friends = {'Name':['Malay','Raghib','Pranjal','Pranav','Raza','Selena','Emma','scarlett'],
           'Marks':[80,85,90,95,87,70,75,80],
           'Gender':['Male','Male','Male','Male','Male','Female','Female','Female']
          }

#here friend is a dictionary , and using this dictionary we created a dataframe.

df = pd.DataFrame(friends)
df

Name  Marks  Gender
0     Malay     80    Male
1    Raghib     85    Male
2   Pranjal     90    Male
3    Pranav     95    Male
4      Raza     87    Male
5    Selena     70  Female
6      Emma     75  Female
7  scarlett     80  Female

In [4]:
#Q1. Display top 3 rows of the Dataset

df.head() #by-default it shows top 5 rows
df.head(3) #prints top 3 rows

Name  Marks Gender
0    Malay     80   Male
1   Raghib     85   Male
2  Pranjal     90   Male

In [5]:
#Q.2 check last 3 Rows of The DataSet

df.tail() #by-defalt it shows 5 last rows

df.tail(5) #displays last 5 rows

Name  Marks  Gender
3    Pranav     95    Male
4      Raza     87    Male
5    Selena     70  Female
6      Emma     75  Female
7  scarlett     80  Female

In [13]:
#Find shape of our Dataset (Number of Rows and Number of columns)
#shape is not a method(funciton),it is an attribute of a panda dataframe

print(df.shape) #prints tuple having no of rows and column in table

print('Number of Rows : ',df.shape[0])
print('Number of columns :',df.shape[1])

(8, 3)
Number of Rows :  8
Number of columns : 3


### Get infromation About Our Dataset like total number of rows and columns, Total number of columns , datatypes of each column and memory rquirement

In [14]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8 entries, 0 to 7
Data columns (total 3 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   Name    8 non-null      object
 1   Marks   8 non-null      int64 
 2   Gender  8 non-null      object
dtypes: int64(1), object(2)
memory usage: 320.0+ bytes


### check Null values in the dataset

In [15]:
df.isnull()

Name  Marks  Gender
0  False  False   False
1  False  False   False
2  False  False   False
3  False  False   False
4  False  False   False
5  False  False   False
6  False  False   False
7  False  False   False

In [20]:
df.isnull().sum() #columnwise

Name      0
Marks     0
Gender    0
dtype: int64

In [19]:
df.isnull().sum(axis = 0) #displays null values column wise

Name      0
Marks     0
Gender    0
dtype: int64

In [21]:
df.isnull().sum(axis = 1) #displayes the null value row wise

0    0
1    0
2    0
3    0
4    0
5    0
6    0
7    0
dtype: int64

# Get overall statistics about the dataframe

In [22]:
df.describe() #it displays on that column that contain numerical value

Marks
count   8.000000
mean   82.750000
std     8.137216
min    70.000000
25%    78.750000
50%    82.500000
75%    87.750000
max    95.000000

In [23]:
df.describe(include = 'all')

Name      Marks Gender
count       8   8.000000      8
unique      8        NaN      2
top     Malay        NaN   Male
freq        1        NaN      5
mean      NaN  82.750000    NaN
std       NaN   8.137216    NaN
min       NaN  70.000000    NaN
25%       NaN  78.750000    NaN
50%       NaN  82.500000    NaN
75%       NaN  87.750000    NaN
max       NaN  95.000000    NaN

 25% marks is below 78.75
 
 
 50% marks is below 82.5
 
 
 75% marks is below 87.75
    
    max is 95

## Find unique values from the Gender column

In [7]:
df

Name  Marks  Gender
0     Malay     80    Male
1    Raghib     85    Male
2   Pranjal     90    Male
3    Pranav     95    Male
4      Raza     87    Male
5    Selena     70  Female
6      Emma     75  Female
7  scarlett     80  Female

In [9]:
df['Gender'].unique()

array(['Male', 'Female'], dtype=object)

## Find the number of unique values from the Gender column

In [10]:
df['Gender'].nunique()

2

## Display count of Unique values in Gender Columnn

In [11]:
df['Gender'].value_counts()

Male      5
Female    3
Name: Gender, dtype: int64

# Find total number of students having marks between 85 to 90(inclusive) using between mehod


In [12]:
df['Marks']

0    80
1    85
2    90
3    95
4    87
5    70
6    75
7    80
Name: Marks, dtype: int64

In [13]:
df['Marks'] >=85

0    False
1     True
2     True
3     True
4     True
5    False
6    False
7    False
Name: Marks, dtype: bool

In [15]:
df[df['Marks']>=85]

Name  Marks Gender
1   Raghib     85   Male
2  Pranjal     90   Male
3   Pranav     95   Male
4     Raza     87   Male

In [22]:
df['Marks'] >= 85 & df['Marks'] <=90 #see the below code and find the difference

ValueError: The truth value of a Series is ambiguous. Use a.empty, a.bool(), a.item(), a.any() or a.all().

In [21]:
#if you see , the code below contains the parenthesis 

(df['Marks'] >= 85) & (df['Marks'] <=90 )


0    False
1     True
2     True
3    False
4     True
5    False
6    False
7    False
Name: Marks, dtype: bool

In [28]:
df[(df['Marks']>=85 )& (df['Marks']<=90)] 

Name  Marks Gender
1   Raghib     85   Male
2  Pranjal     90   Male
4     Raza     87   Male

In [29]:
len(df[(df['Marks']>=85 )& (df['Marks']<=90)])

3

In [31]:
df['Marks'].between(85,100) #similar to x >= 85 and x<=90 

0    False
1     True
2     True
3     True
4     True
5    False
6    False
7    False
Name: Marks, dtype: bool

here in above code there are 4 true value and 4 false value , True means 1. 
thus sum() returns 4.

In [32]:
sum(df['Marks'].between(85,100)) 

4

# Find Average Marks

In [5]:
df

Name  Marks  Gender
0     Malay     80    Male
1    Raghib     85    Male
2   Pranjal     90    Male
3    Pranav     95    Male
4      Raza     87    Male
5    Selena     70  Female
6      Emma     75  Female
7  scarlett     80  Female

In [6]:
df['Marks'].mean()

82.75

In [7]:
df['Marks'].min()

70

In [8]:
df['Marks'].max()

95

# APPLY METHOD

In [14]:
def marks(x):
    return x/2 #use x//2 for floor division

In [15]:
df['Marks'].apply(marks)

0    40.0
1    42.5
2    45.0
3    47.5
4    43.5
5    35.0
6    37.5
7    40.0
Name: Marks, dtype: float64

In [17]:
df['Half-Marks'] = df['Marks'].apply(marks)

In [18]:
df

Name  Marks  Gender  Half-Marks
0     Malay     80    Male        40.0
1    Raghib     85    Male        42.5
2   Pranjal     90    Male        45.0
3    Pranav     95    Male        47.5
4      Raza     87    Male        43.5
5    Selena     70  Female        35.0
6      Emma     75  Female        37.5
7  scarlett     80  Female        40.0

In [20]:
df['Marks'].apply(lambda x:x/2)

0    40.0
1    42.5
2    45.0
3    47.5
4    43.5
5    35.0
6    37.5
7    40.0
Name: Marks, dtype: float64

a temporary anonymous function is made in .apply itself using lambda

In [21]:
#Print length of string in Name column
df['Name'].apply(len)

0    5
1    6
2    7
3    6
4    4
5    6
6    4
7    8
Name: Name, dtype: int64

In [24]:
df['Name'].apply(lambda x:len(x))

0    5
1    6
2    7
3    6
4    4
5    6
6    4
7    8
Name: Name, dtype: int64

# Map Funciton

In [25]:
df

Name  Marks  Gender  Half-Marks
0     Malay     80    Male        40.0
1    Raghib     85    Male        42.5
2   Pranjal     90    Male        45.0
3    Pranav     95    Male        47.5
4      Raza     87    Male        43.5
5    Selena     70  Female        35.0
6      Emma     75  Female        37.5
7  scarlett     80  Female        40.0

Target : Convert Male to 1 and Female to 0

In [26]:
df['Gender'].map({'Male':1,'Female':0})

0    1
1    1
2    1
3    1
4    1
5    0
6    0
7    0
Name: Gender, dtype: int64

In [27]:
df

Name  Marks  Gender  Half-Marks
0     Malay     80    Male        40.0
1    Raghib     85    Male        42.5
2   Pranjal     90    Male        45.0
3    Pranav     95    Male        47.5
4      Raza     87    Male        43.5
5    Selena     70  Female        35.0
6      Emma     75  Female        37.5
7  scarlett     80  Female        40.0

In [29]:
df['Male_Female'] = df['Gender'].map({'Male':1,'Female':0}) #adds a new column Male_Femaale

In [30]:
df

Name  Marks  Gender  Half-Marks  Male_Female
0     Malay     80    Male        40.0            1
1    Raghib     85    Male        42.5            1
2   Pranjal     90    Male        45.0            1
3    Pranav     95    Male        47.5            1
4      Raza     87    Male        43.5            1
5    Selena     70  Female        35.0            0
6      Emma     75  Female        37.5            0
7  scarlett     80  Female        40.0            0

# Drop the columns

In [33]:
df.drop('Half-Marks',axis = 1) #axis = 1 indicate the column in dataframe

Name  Marks  Gender  Male_Female
0     Malay     80    Male            1
1    Raghib     85    Male            1
2   Pranjal     90    Male            1
3    Pranav     95    Male            1
4      Raza     87    Male            1
5    Selena     70  Female            0
6      Emma     75  Female            0
7  scarlett     80  Female            0

In [36]:
df.drop('Male_Female')

KeyError: "['Male_Female'] not found in axis"

In [37]:
df.drop('Male_Female',axis = 1)

Name  Marks  Gender  Half-Marks
0     Malay     80    Male        40.0
1    Raghib     85    Male        42.5
2   Pranjal     90    Male        45.0
3    Pranav     95    Male        47.5
4      Raza     87    Male        43.5
5    Selena     70  Female        35.0
6      Emma     75  Female        37.5
7  scarlett     80  Female        40.0

In [ ]:
#droping Multiple columns

In [38]:
df.drop(['Male_Female','Half-Marks'],axis = 1)

Name  Marks  Gender
0     Malay     80    Male
1    Raghib     85    Male
2   Pranjal     90    Male
3    Pranav     95    Male
4      Raza     87    Male
5    Selena     70  Female
6      Emma     75  Female
7  scarlett     80  Female

In [39]:
df

Name  Marks  Gender  Half-Marks  Male_Female
0     Malay     80    Male        40.0            1
1    Raghib     85    Male        42.5            1
2   Pranjal     90    Male        45.0            1
3    Pranav     95    Male        47.5            1
4      Raza     87    Male        43.5            1
5    Selena     70  Female        35.0            0
6      Emma     75  Female        37.5            0
7  scarlett     80  Female        40.0            0

In [ ]:
#Now if you want to modify and update the table. use

#inplace = True

In [41]:
df.drop(['Half-Marks','Male_Female'],axis = 1, inplace = True)

In [42]:
df

Name  Marks  Gender
0     Malay     80    Male
1    Raghib     85    Male
2   Pranjal     90    Male
3    Pranav     95    Male
4      Raza     87    Male
5    Selena     70  Female
6      Emma     75  Female
7  scarlett     80  Female

In [ ]:
Print Name of the Columns

In [43]:
df.columns

Index(['Name', 'Marks', 'Gender'], dtype='object')

In [44]:
df.index #prints range(start,end,step)

RangeIndex(start=0, stop=8, step=1)

# Sort the dataFrame as per the Marks Column

In [45]:
df

Name  Marks  Gender
0     Malay     80    Male
1    Raghib     85    Male
2   Pranjal     90    Male
3    Pranav     95    Male
4      Raza     87    Male
5    Selena     70  Female
6      Emma     75  Female
7  scarlett     80  Female

In [49]:
df.sort_values(by='Marks',ascending= False)  #shift + tab to know docstring

Name  Marks  Gender
3    Pranav     95    Male
2   Pranjal     90    Male
4      Raza     87    Male
1    Raghib     85    Male
0     Malay     80    Male
7  scarlett     80  Female
6      Emma     75  Female
5    Selena     70  Female

In [54]:
df.sort_values(by = ['Marks','Gender'], ascending=False)

Name  Marks  Gender
3    Pranav     95    Male
2   Pranjal     90    Male
4      Raza     87    Male
1    Raghib     85    Male
0     Malay     80    Male
7  scarlett     80  Female
6      Emma     75  Female
5    Selena     70  Female

# Adding a new Row in dataframe

In [56]:
df1 = {'Name':'Manas','Marks': 96,'Gender':'Trans'}
df = df.append(df1,ignore_index = True)
display(df)

Name  Marks  Gender
0     Malay     80    Male
1    Raghib     85    Male
2   Pranjal     90    Male
3    Pranav     95    Male
4      Raza     87    Male
5    Selena     70  Female
6      Emma     75  Female
7  scarlett     80  Female
8     Manas     96   Trans
9     Manas     96   Trans

In [57]:
df

Name  Marks  Gender
0     Malay     80    Male
1    Raghib     85    Male
2   Pranjal     90    Male
3    Pranav     95    Male
4      Raza     87    Male
5    Selena     70  Female
6      Emma     75  Female
7  scarlett     80  Female
8     Manas     96   Trans
9     Manas     96   Trans

# Deleting a row of dataframe

In [60]:
df.drop(9, axis = 0 , inplace = True)

In [61]:
df.sort_values(['Marks','Gender'], ascending = False)

Name  Marks  Gender
8     Manas     96   Trans
3    Pranav     95    Male
2   Pranjal     90    Male
4      Raza     87    Male
1    Raghib     85    Male
0     Malay     80    Male
7  scarlett     80  Female
6      Emma     75  Female
5    Selena     70  Female

# Update the value in a Row

In [64]:
df.at[8,'Marks'] = 94

In [65]:
df

Name  Marks  Gender
0     Malay     80    Male
1    Raghib     85    Male
2   Pranjal     90    Male
3    Pranav     95    Male
4      Raza     87    Male
5    Selena     70  Female
6      Emma     75  Female
7  scarlett     80  Female
8     Manas     94   Trans

In [67]:
df.sort_values(['Marks','Gender'] ,ascending = False)

Name  Marks  Gender
3    Pranav     95    Male
8     Manas     94   Trans
2   Pranjal     90    Male
4      Raza     87    Male
1    Raghib     85    Male
0     Malay     80    Male
7  scarlett     80  Female
6      Emma     75  Female
5    Selena     70  Female

In [68]:
df.sort_values(['Gender','Marks'] ,ascending = False) 

Name  Marks  Gender
8     Manas     94   Trans
3    Pranav     95    Male
2   Pranjal     90    Male
4      Raza     87    Male
1    Raghib     85    Male
0     Malay     80    Male
7  scarlett     80  Female
6      Emma     75  Female
5    Selena     70  Female

# Display Name and Marks of the female students

In [69]:
df

Name  Marks  Gender
0     Malay     80    Male
1    Raghib     85    Male
2   Pranjal     90    Male
3    Pranav     95    Male
4      Raza     87    Male
5    Selena     70  Female
6      Emma     75  Female
7  scarlett     80  Female
8     Manas     94   Trans

In [70]:
df['Gender']

0      Male
1      Male
2      Male
3      Male
4      Male
5    Female
6    Female
7    Female
8     Trans
Name: Gender, dtype: object

In [73]:
df[df['Gender'] == 'Female']

Name  Marks  Gender
5    Selena     70  Female
6      Emma     75  Female
7  scarlett     80  Female

But we want only Name and Marks

In [79]:
df[df['Gender']== 'Female'] [['Name','Marks']]

Name  Marks
5    Selena     70
6      Emma     75
7  scarlett     80

In [83]:
#Method 2
df['Gender'].isin(['Female'])

0    False
1    False
2    False
3    False
4    False
5     True
6     True
7     True
8    False
Name: Gender, dtype: bool

In [81]:
df[df['Gender'].isin(['Female'])] 

Name  Marks  Gender
5    Selena     70  Female
6      Emma     75  Female
7  scarlett     80  Female

In [84]:
df[df['Gender'].isin(['Female'])][['Name','Marks']]

Name  Marks
5    Selena     70
6      Emma     75
7  scarlett     80